contrib/inventory/foreman.py

```diff
@@ -0,0 +1,391 @@
+#!/usr/bin/env python
+# vim: set fileencoding=utf-8 :
+#
+# Copyright (C) 2016 Guido Günther <agx@sigxcpu.org>
+#
+# This script is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with it.  If not, see <http://www.gnu.org/licenses/>.
+#
+# This is somewhat based on cobbler inventory
+
+from __future__ import print_function
+
+import argparse
+import copy
+import os
+import re
+import requests
+from requests.auth import HTTPBasicAuth
+import sys
+from time import time
+
+try:
+    import ConfigParser
+except ImportError:
+    import configparser as ConfigParser
+
+
+try:
+    import json
+except ImportError:
+    import simplejson as json
+
+
+class ForemanInventory(object):
+    config_paths = [
+        "/etc/ansible/foreman.ini",
+        os.path.dirname(os.path.realpath(__file__)) + '/foreman.ini',
+    ]
+
+    def __init__(self):
+        self.inventory = dict()  # A list of groups and the hosts in that group
+        self.cache = dict()   # Details about hosts in the inventory
+        self.params = dict()  # Params of each host
+        self.facts = dict()   # Facts of each host
+        self.hostgroups = dict()  # host groups
+
+    def run(self):
```

Comment:
Typically we organize code so that callers are closer to the bottom and callees are near the top (```__init__``` being the exception... it's always at the top).  This organization puts methods like ```run()``` near the bottom of the class and helpers above it.